backend/test_observer/controllers/artefacts/artefacts.py

```diff
@@ -70,14 +74,36 @@ def get_artefact(artefact_id: int, db: Session = Depends(get_db)):
 def patch_artefact(
     artefact_id: int, request: ArtefactPatch, db: Session = Depends(get_db)
 ):
-    artefact = db.get(Artefact, artefact_id)
-
+    query_options = []
+    if request.status in {ArtefactStatus.APPROVED, ArtefactStatus.MARKED_AS_FAILED}:
+        # Load test executions as we need to check them
+        query_options.append(
+            joinedload(Artefact.builds).joinedload(ArtefactBuild.test_executions),
+        )
+    artefact = db.get(Artefact, artefact_id, options=query_options)
     if artefact is None:
         raise HTTPException(status_code=404, detail="Artefact not found")
 
+    if (
+        request.status == ArtefactStatus.APPROVED
+        and not are_all_test_executions_approved(artefact)
+    ):
+        raise HTTPException(
+            status_code=400,
+            detail="All test executions need to be approved",
+        )
+
+    if (
+        request.status == ArtefactStatus.MARKED_AS_FAILED
+        and not is_there_a_rejected_test_execution(artefact)
+    ):
+        raise HTTPException(
+            400,
+            detail="At least one test execution needs to be rejected",
+        )
```

Comment:
Just a small code style suggestion, maybe we can extract this into helper `_validate_request_status` to clear up the main login of this method.